<a href="https://colab.research.google.com/github/LhermannSauer/dm2025a/blob/main/src/CazaTalentos/a_cazatalentos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Cazatalentos
## buscando a  “La verdadera mejor encestadora”

## 1.Introduccion
Este notebook analiza de forma didáctica la razón por la que se produce el overfitting.

Ver los cinco minutos de este video muy relacionado con lo que veremos
[Veritasium El Éxito ¿es Suerte o Trabajo Duro?](https://storage.googleapis.com/open-courses/austral2025-af91/labo1r/veritasium_success_paradox.html)

### 1.1 Motivacion

El siguiente desafío busca introducir el problema fundamental de la Ciencia de Datos que es el overfitting desde un ángulo distinto,  presentando los fenómenos del efecto del tamaño de la muestra,
<br> el conocido problema de las múltiples comparaciones  y “la maldición del ganador”.
<br>
<br> Aunque anecdóticametne está basado en algo deportivo, se extiende fácilmente  a :
*  la efectividad de acciones de marketing
*  el testeo de medicamentos
*  experimentos ciéntíficos https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1182327/
*  la comparación de modelos predictivos
*  la elección del mejor corte en un nodo de un árbol de decisión, etc

### 1.2 Generalidades
Vari@s cazatalent@s de la gran ciudad van recorriendo pueblos en búsqueda de las mejores jugadoras de basket para tiros libres en la Liga Femenina de Basket.
<br>El procedimiento recomendado es el siguiente : un@ cazatalentos llega a un pueblo y va a la única cancha de basket del lugar donde la están esperando las adolescentes interesadas, las hace tirar tiros libres y elige a la mejor jugadora.
<br>Posteriormente a la elección del mejor jugadora, si le parece a la entrenadora de la gran ciudad, se llevará a la jugadora a que demuestre que tan buena es.

<br>A las jugadoras los podemos equiparar con modelos predictivos.


### 2 Definición del Problema

¿Qué metodología de las que se plantearán es la que propone a la jugadora con mayor probabilidad de superar el umbral del 80% de aciertos en la nueva prueba que se hará en la gran ciudad?
<br> Cada jugadora posee un “índice de enceste”, la cazatalentos no tiene forma de conocerlos; lo único que puede hacer la cazatalentos es hacer que las jugadoras hagan tiros libres, y medir los aciertos y no aciertos.

### 2.1  Caso fallido, una jugadora se quiere pasar de lista ...

La cazatalentos llega a una pequeña localidad a la que le han dicho que hay dos adolescentes cuyas habilidades debe testear.
<br> Una es super talentosa se llama Diana Taurasi  https://en.wikipedia.org/wiki/Diana_Taurasi, nosotros sabemos que tiene un “índice de enceste” del 0.85, es decir cada vez que hace un tiro libre su probabilidad de encestar es del 85%.   Si hace 100 tiros libres, en promedio encesta 85 veces,  a veces puede ser más, a veces menos, pero en promedio encesta 85 de 100.
<br> La otra es un auténtico desastre, a quien llamaremos  Miss X, y su índice de enceste es de 0.10.  Esto quiere decir que si hace 100  tiros libres, en promedio encesta 10.
<br> La cazatalentos NO tiene forma de conocer los valores de 0.85 y 0.10, solo puede hacerlas tirar y contar aciertos.

<br>La cazatalentos llega al gimnasio,  y solo se encuentra Diana, con lo cual le indica que comience a tirar.  Al tiro 10, Diana lleva 8 encestes y 2 no-enceste.
<br>En ese momento, llega con un andar cansino Miss X y las interrumpe :  ¿cuánto va Diana ? La cazatalentos le responde  8 de 10.  Miss X solicita permiso para tirar, tira una sola vez y acierta.
<br>Ahí mismo dice : “paren las rotativas, yo soy la clara ganadora de este torneo,  mi indice de acierto es del  100% y el de Diana es del 80% ,  la matemática es clara”.
<br>La cazatalentos monta en cólera y le dice :  pues usted joven solo ha realizado un tiro, ¿Si lo hubiera errado, aceptaría decir entonces que su indice de acierto es del 0%? Ciertamente no.
<br>Pues ahora, usted deberá seguir tirando hasta completar sus 10 tiros libres, y recién ahí compararé.


### 2.2 Segundo Caso

La cazatalentos tiene el siguiente algoritmo, hace que Diana Taurasi y Miss X hagan 10 tiros libres cada una, y elige a la ganadora de esos 10 tiros libres.
¿Garantiza este método que por lo menos el 99% de las veces elige a Diana Taurasi?

<br>¿La desastrosa Miss X puede llegar a encestar los 10 tiros libres?  <br>Efectivamente sí existe esa posibilidad, pero con una muy baja probabilidad, la probabilidad es de 0.10 ^10  =  1e-10
<br>El símbolo de  ^  significa  "elevado a la potencia de"
<br>
<br>¿Cual es la probabilidad que Miss X tire una sola vez y acierte este tiro? <br>Simplemente 0.1 , una en diez, con lo cual ha tenido mucha suerte en su primer tiro, pero no la va a poder mantener !

<br>¿La adolescente Diana Taurasi puede llegar a encestar los 10 tiros libres?  Efectivamente, su probabilidad es  0.85 ^10 = 0.1968744

En el lenguaje R la instrucción  runif(10)  genera un vector con 10 números aleatorios con distribución uniforme en el intervalo [0,1], por ejemplo

In [ ]:
rand(10)

10-element Vector{Float64}:
 0.23525012031240244
 0.5580344315340746
 0.9215912506606294
 0.9850399754219642
 0.48257397476296526
 0.6895107126216442
 0.910898693697144
 0.16027279605493905
 0.3877978829076234
 0.4335225409105655

Ahora simulamos 10000 veces la estrategia de tirar 10 tiros libres y quedarse con la ganadora.

In [ ]:
missx_ganadora = 0

for i  = 1:10000
  aciertos_taurasi = sum( rand(10) .< 0.85 )
  aciertos_missx = sum( rand(10) .< 0.10 )

  if( missx_ganadora > aciertos_missx )  missx_ganadora <- missx_ganadora +1 end
end
println( "Miss X gano un total de : $missx_ganadora" )

Miss X gano un total de : 0


Lo que da la cantidad de 0,  o sea que en 10000 veces, siempre ganó Diana Taurasi, con lo cual la estrategia de quedarse con el mejor de 10 tiros ha funcionado, para este obvio caso donde hay apenas dos jugadoras y hay una diferencia abismal entre ellas.

En este caso NO aparece el overfitting por ningún lado

### 2.3  Aparece el Overfitting

El cazatalentos llega ahora a un pueblo donde hay 100 jugadores los que son mucho más parejos entre sí desde el punto de vista del ratio de enceste.

Hay un jugador, que llamaremos jugador 1 que tiene un “índice de enceste” de  0.70
Los 99 jugadores restantes que llamaremos “el pelotón”  tienen los índices de enceste  { 0.501, 0.0502, 0.503, …, 0.599  }
Es decir el jugador 1 tiene 0.70 y el mejor jugador del pelotón tiene 0.599,  hay un poco más de 0.10 de diferencia.  La diferencia de 0.10 es significativa en basket.

¿Qué sucede con el algoritmo del cazatalentos de hacer tirar 10 tiros libres a cada uno de los jugadores y elegir al que más encestes logró ?


In [ ]:
using Random

Random.seed!(102191)

# calcula cuantos encestes logra una jugadora con indice de enceste prob
# haciendo qyt tiros libres

ftirar = (prob, q) -> (sum(rand(q) .< prob)/ q)


# defino las jugadoras
taurasi    = 0.7
peloton    = .501:0.001:.599
jugadoras  = cat( taurasi, peloton,dims=(1))





# hago que las 100 jugadoras tiren 10 veces cada una
primera_ganadora  = 0

for i = 1:10000  # diez mil experimentos

  vaciertos  = ftirar.(jugadoras, 10)  # 10 tiros libres cada jugadora

  mejor  = findmax( vaciertos )[2]
  if mejor == 1  primera_ganadora  = primera_ganadora + 1 end
end

println( "En los diez mil experimentos, se detecto correctamente a la mejor jugadora la cantidad de veces: $primera_ganadora ")


En los diez mil experimentos, se detecto correctamente a la mejor jugadora la cantidad de veces: 1363 


Aquí ya pasa algo asombroso, el overfitting en todo su esplendor.
Si hago tirar 10 tiros libres a cada uno de las 100 jugadoras, apenas 1450 veces de las 10000 este método  devuelve a la verdadera mejor, que es la jugadora con un índice de enceste de 0.70

Intuitivamente, lo que está sucediendo es que la inmensa mayoría de las veces, una de las 99 jugadoras del pelotón tiene mucha suerte y supera a la jugadora de 0.70, con lo cual se elige a una del pelotón y no a la verdadera mejor !  
<br>El tener 99 jugadoras en el pelotón hace que al ser tantas hay varias que tienen mala suerte, pero también hay varias con muy buena suerte, y la que tuvo más suerte supera a la verdadera mejor !


### 2.4 Nueva ronda

Ahora pasemos a ver aún algo más notable.
Tenemos a nuestra jugadora de 0.70, y a las 99 jugadoras del pelotón.
1.  Las hacemos tirar a cada una 10 tiros libres, y elegimos a la ganadora registrando cuántos aciertos tuvo
2.  solo a esa ganadora la hacemos tirar una nueva ronda de 10 tiros libres
3.  finalmente comparamos estos nuevos aciertos con los originales.


In [ ]:
Random.seed!(102191)
println( "Aciertos Ganadora", "\t", "Nueva Ronda", "\n")

for i in 1:10
  vaciertos  = ftirar.(jugadoras, 10 )  # cada jugadora tira 10 tiros libres
  _, mejor  = findmax( vaciertos )
  aciertos_torneo  = vaciertos[ mejor ]

  aciertos_segunda  = ftirar.( jugadoras[ mejor ], 10 )

  println( aciertos_torneo, "\t\t\t", aciertos_segunda, "\n" )
end

Aciertos Ganadora	Nueva Ronda

0.9			0.7

0.9			0.6

0.9			0.6

0.8			0.4

0.9			0.5

0.9			0.7

1.0			0.5

0.9			0.7

0.9			0.3

0.9			0.7



Lo que se observa en este caso es “La maldición del ganador”,
<br>la performance que la jugadora que logró más aciertos en la competencia general luego NO LA PUEDE MANTENER,
<br>en todos los casos vemos que en la nueva ronda de diez tiene menos aciertos que en la primera ronda de diez.
<br>La primera ronda  la jugadora que resultó ganadora fue debido a la suerte, y esa suerte ya no la acompaña para la segunda ronda de 10 tiros libres.

Relacionado con las competencias Kaggle, una forma de pensar esto es  "¿Cómo le va a ir al mejor modelo del Public Leaderboard en el Private?" ,  esta es la razón por la que los puntajes que vemos en el Private son inferiores al Public.
La primer ronda puede verse como quedarnos con el mejor modelo del Public Leaderboard, la nueva ronda es como le va a ese modelo en el Private Leaderboard.
<br>Elegir el mejor modelo del Public, suele ser una pésima estrategia.

### El overfitting en su plenitud

¿Cuándo es más extremo el efecto de “La maldición del ganador” ?
<br>Supongamos por un momento que ahora, en otro pueblo,  tenemos 100 jugadoras nuevas, todas con un “índice de enceste” de 0.70 ;  recordar que ese valor jamás es conocido por la cazatalentos.

<br>Ahora nos ponemos más estrictos, y las hacemos tirar 100 veces a cada una, elegimos a la ganadora, y solo a esa ganadora la hacemos tirar 100 nuevos tiros libres.
<br> Podemos pensarlo como que la cazatalentos se lleva a la mejor jugadora a la gran ciudad, le habla maravillas de ella a la entrenadora, y la entrenadora dice “probémosla a ver si es tan buena como decís, que haga 100 tiros libres aquí y ahora”

In [ ]:
Random.seed!()
jugadoras  = repeat([0.7], 100 )

for i in 1:10
  vaciertos  = ftirar.(jugadoras, 100 )  # cada jugadora tira 10 tiros libres
  _, mejor  = findmax( vaciertos )
  aciertos_torneo  = vaciertos[ mejor ]

  aciertos_segunda  = ftirar.( jugadoras[ mejor ], 10 )

  println( aciertos_torneo, "\t\t\t", aciertos_segunda, "\n" )
end



0.82			0.8

0.8			0.6

0.83			0.8

0.8			0.6

0.78			0.6

0.82			0.6

0.79			0.7

0.82			0.5

0.81			0.6

0.81			0.9



Nuevamente vemos, que el puntaje alcanzado por la mejor jugadora cuando compitió contra las otras 99,  NO ES VUELTO A ALCANZAR  en la nueva ronda de 100 tiros libres que hace esa jugadora.
<br>Este efecto es exactamente el mismo que vemos cuando alguien se empecina en lograr en el Leaderboard Público el mayor puntaje de la clase,  le va muy bien en el Público, pero se derrumba catastróficamente en el Privado.
<br>Decimos que  “overfitió / sobreajustó el Leaderboard Publico”.

### 2.5 medida de la diferencia

Ahora nos preguntamos, en este caso, en promedio, cuanto más aciertos tiene la ganadora (debido a la suerte)  en la primera ronda  con respecto a la segunda ?


In [ ]:


Random.seed!( )

# defino las jugadoras
jugadoras  = repeat( [0.7], 100 )

suma_diferencias = 0

for i = 1:10000
  vaciertos  = ftirar.(jugadoras, 100 )  # cada jugador tira 100 tiros libres
  _, mejor  = findmax( vaciertos )
  aciertos_torneo  = vaciertos[ mejor ]

  aciertos_segunda  = ftirar.(jugadoras[mejor], 100 )

  suma_diferencias  = suma_diferencias +  (aciertos_torneo - aciertos_segunda )
end


println( "Diferencia promedio: $(suma_diferencias / 10000)" )


Diferencia promedio: 0.02574800000000037


Exactamente este mismo efecto se da en el marketing digital.  Si en un A/B testing se prueban cien alternativas y se elige a la que tiene más efectividad, se observará que en un nuevo experimento disminuirá el rendimiento de esa alternativa.

También sucede si se prueban muchos medicamentos al mismo tiempo,   el que resulte más efectivo disminuirá su efectividad en la nueva prueba.

Este problema es llamado “El problema de las múltiples comparaciones”

### 2.6 una sola jugadora MUY buena

Sexto caso,  ¿cuándo se atenúa/desaparece  el overfitting?
Supongamos 99 jugadoras con un índice de enceste de 0.60   y  ahora sumamos a la adolescente Diana Taurasi con su superlativo índice de enceste del 0.85
Los hacemos tirar a todas 100 tiros libres, llevamos a la ganadora a la gran ciudad, y allí le contamos a la entrenadora de nuestra ganadora Diana Taurasi.

In [ ]:


Random.seed!( 102191 )

# defino las jugadoras
taurasi    = 0.85
peloton   = repeat( [.6], 99 )  # jugadoras identicas
jugadoras = vcat( [taurasi], peloton )



println( "Aciertos Ganadora", "\t", "Nueva Ronda", "\n")

for i = 1:10
  vaciertos  = ftirar.(jugadoras, 100 )  # cada jugadora tira 100 tiros libres
  _, mejor  = findmax( vaciertos )
  aciertos_torneo  = vaciertos[ mejor ]

  aciertos_segunda  = ftirar( jugadoras[ mejor ], 100 )

  println(aciertos_torneo, "\t\t\t", aciertos_segunda, "\n" )
end



Aciertos Ganadora	Nueva Ronda

0.85			0.9

0.84			0.84

0.83			0.88

0.85			0.8

0.87			0.85

0.87			0.86

0.85			0.86

0.8			0.82

0.81			0.84

0.88			0.88



En este caso observamos dos cosas, en primer lugar SIEMPRE la ganadora del torneo fue Diana Taurasi, y en segundo  cuando tuvo que mostrar en la ciudad su performance, fue la misma que en el primer torneo.  
<br>Es tan buena Taurasi que le ganó a la que tuvo más suerte del pelotón, el efecto suerte no está afectando a Taurasi porque está sola, y gana por su superioridad, no por la suerte.

### 2.7 Efectos del tamaño de la ronda

La cazatalentos llega ahora a un pueblo donde hay 100 jugadoras las que son mucho más parejas entre sí desde el punto de vista del ratio de enceste.
Hay un jugadora, que llamaremos jugadora 1 que tiene un “índice de enceste” de  0.70 ,  es una jugadora buena pero dista mucho de ser Diana Taurasi, jamás estará en el "WNBA Hall of Fame".

Las 99 jugadoras restantes que llamaremos “el pelotón”  tienen los índices de enceste  { 0.501, 0.0502, 0.503, …, 0.599  }  Es decir la jugadora 1 tiene 0.70 y la mejor jugadora del pelotón tiene 0.599,  o sea hay un poco más de 0.10 de diferencia.  La diferencia de 0.10 es significativa en basket.

¿Cuántos tiros libres debe la cazatalentos pedirles que haga cada jugadora para que  si elige a la mejor jugadora tiene la certeza de llevarse a “la verdadero mejor” el 99% de los casos ?


In [ ]:
Random.seed!(102191)
ftirar = (prob, q) -> (sum(rand(q) .< prob)/ q)


# # defino los jugadoras
taurasi = .7
peloton = .501:.001:0.599
jugadoras = cat(peloton, taurasi, dims=(1))

print("Tiros \t\t Taurasi \n")


for tiro_libre in [10, 20, 50, 100, 200, 300, 400, 415, 500, 600, 700, 1000]
  ultima_ganadora = 0
  for i = 1:10000
    vaciertos = ftirar.(jugadoras, tiro_libre)
    _, mejor = findmax(vaciertos)
    if mejor == 100
      ultima_ganadora = ultima_ganadora + 1
    end
  end

  print("$tiro_libre \t\t $(ultima_ganadora/100) \n")

end




Tiros 		 Taurasi 
10 		 3.2 
20 		 8.75 
50 		 27.48 
100 		 55.65 
200 		 84.98 
300 		 96.02 
400 		 98.96 
415 		 99.0 
500 		 99.66 
600 		 99.96 
700 		 99.97 
1000 		 100.0 


Haciendo tirar 415 tiros libres a cada una de los 100 jugadoras,
<br> y quedándose con la ganadora de ese torneo,
<br> en el 99.10%   ( 0.9910 )  de los casos se está segura que esa es "la verdadera mejor".


## 3 Conclusiones

El overfitting aparece cuando se comparan muchas jugadoras (modelos predictivos)  muy parecidas entre sí.  
<br> En el modelado predictivo es la situación más común que se tengan modelos con similar poder predictivo ya que generalmente solo difieren apenas en valores de los hiperparámetros o algunas columnas nuevas con feature engineering.

Las jugadoras no son determinísticas, sino probabilísticas, con lo cual al comparar muchas parecidas la ganadora es quien además de su talento tuvo más suerte en ese torneo.  La jugadora ganadora ganó por mera suerte, y  NO puede sostener esa suerte en un nuevo torneo, por lo que el puntaje del primer torneo es mentiroso, no se sostiene en el nuevo torneo.

El overfitting se atenúa o puede llegar a desaparecer del todo, si en el torneo hay una jugadora ampliamente superior al resto. Esa jugadora SI mantiene su performance de un torneo al otro, ya que su puntaje en el primer torneo fue alto porque es realmente buena, porque no es el máximo de muchos parecidas a ella.

La forma de combatir el overfitting es hacer torneos de muchos tiros libres.
<br>Para estimar que tan buena es la jugadora ganadora de un torneo, hay que hacerla tirar nuevamente tiros libres.
<br>El secreto está en hacer más de un torneo.


## 4 Desafio Cazatalentos Ordenamiento

### 4.1 Enunciado

Se reúne la entrenadora  en la gran ciudad con algunas cazatalentos que visitaron regiones distintas.

Entrenadora:  Bienvenidas a todas, veo que cada una de ustedes ha traído la estadística de su candidata. Sin embargo, yo debo ordenarlas de mejor a peor, y no me alcanza con que dos de ustedes me digan que que sus candidatas encestaron 80/100, necesito que me relaten exactamente el procedimiento que han seguido para elegir a la mejor jugadora.

Cazatalentos 1:  Discúlpeme entrenadora, si dos de nosotras le presentamos casos que en distintos pueblos han encestado exactamente 80 de 100, probabilisticamente esas adolescentes son equivalentes. Es cierto que luego aquí en la gran ciudad el dia de la prueba una podrá encestar 77 de 100, la otra 83,  pero eso es la parte natural del random, ya sabemos que la distribución es binomial.  Ambas son equivalentes, no importa que pasó con el resto de las jugadoras de su pueblo.

Entrenadora: Estimada “Cazatalentos 1”  tú estas equivocada, no consideras El Problema de las Multiples Comparaciones; elegir cuales de esas adolescentes vendrán a la gran ciudad a ser evaluadas es mi decisión. Comienza tú contando detalladamente el proceso que has seguido.
4





```
# This is formatted as code
```

**Cazatalentos 1**: Llegué a la localidad habían 100 adolescentes en el gimnasio, hice tirar a cada una 100 tiros, solo una obtuvo 80 aciertos, luego un par 79, otro par 78, y así descendían. La que logró 80 aciertos es mi candidata.


**Cazatalentos 2:** Llegué a una pequeña ciudad y habían 200 adolescentes en el gimnasio, hice tirar a cada una 100 tiros, solo una obtuvo 80 aciertos, seis 79,  cinco 78, y así descendían. La que logró 80 aciertos es mi candidata.

**Cazatalentos 3:** Yo poseo 25 años de experiencia como cazatalentos y tengo muy buena intuición  para detectar el talento. Llegué a un pueblo donde habían 100 adolescentes, me senté en las gradas del gimnasio y las observé detenidamente mientras precalentaban sin tirar al aro.  A los 10 minutos, me paro, y señalo a una de ellas y le digo :  'tú, la número 13, a tirar 100 tiros; el resto gracias por haber venido pero se pueden ir cada una a su casa”. Y ahí al tirar esos 100 tiros logró 80 aciertos y esa es mi candidata.

**Cazatalentos 4:** Llegué a minúsculo pueblo con un aro reglamentario en la pared de una escuela, habían solamente 2 adolescentes, hice tirar a cada una 100 tiros, una obtuvo 80 y la otra 75.

**Cazatalentos 5:** Le dice a la entrenadora : “fui a una localidad de 500 habitantes, había una sola adolescente interesada, la hice tirar 100 tiros libres, y encestó 80,  tengo una filmación con mi celular de los 100 tiros, sin edición ni cortes”
En realidad, esa adolescente es la sobrina de la cazatalentos, la hizo tirar 10 rondas de 100 tiros cada una, grabó todas, pero tramposamente solo trae la grabación de la mejor ronda a la entrenadora.
Los resultados obtenidos por su sobrina son


| ronda | Tiros | Encestes |
| ---: | ---: | ---: |
| 1 | 100 | 68 |
| 2 | 100 | 74 |
| 3 | 100 | 78 |
| 4 | 100 | 70 |
| 5 | 100 | 68 |
| 6 | 100 | 63 |
| 7 | 100 | 80 |
| 8 | 100 | 68 |
| 9 | 100 | 67 |
| 10 | 100 | 65 |


(esta deshonestidad de la Cazatalentos 5 no debería parecerle extraña ni causarle gracia, es exactamente lo mismo que elegir como submit final el que le fue mejor en el Public Leaderboard,  algo muy común entre l@s alumn@s ...)
<br>Por una persona espía en el pueblo, la entrenadora está al tanto de lo ocurrido y conoce la tabla anterior, usará dicha información para tomar la decisión, aunque no le dirá nada, por esta vez, a la cazatalentos.

70.1

### 4.2 Desafio

Podemos suponer que la supertalentosa Diana Taurasi con su “índice de enceste” del 0.85 es superior a todas las jugadoras evaluadas en los pueblos por las cazatalentos.

Desafío : ¿Si usted fuera la entrenadora, como ordena el valor esperado de encestes en una ronda de 100 tiros libres  a cada una de las candidatas propuestas por las cinco cazatalentos ?


donde   Ci  < Cj   sii   prob(  #Encestes Ci  <  #Encestes Cj    | 100 tiros cada uno )  >  0.5
es decir que, haciendo 100 tiros libres,  más de la mitad de las veces,  Ci va a lograr menos encestes que Cj

La entrenadora de la gran ciudad NO quiere traer a la elegida de cada cazatalentos y hacerlas tirar 100 tiros delante de ella para constatar cual logra más encestes en esa ronda. Ella quiere, con la información que le brindó cada cazatalentos sobre el procedimiento y los resultados, elegir la mejor encestadora.


### 4.3  Reglas del Desafio


*  La fecha de finalización del desafío es el sabado 02 de agosto a las 19:00
Usted podrá trabajar en forma individual o formando grupo de dos personas.
*  Entrega
      * la entrega se realiza con un solo mensaje en Zulip canal #CazaTalentos
         * Ordenamiento,  por ejemplo     C3 <  C5 <  C1 < C4 < C2
         * Explicacion
      * Si necesitará apoyarse en scripts para alguna de las comparaciones
          * link a carpeta en GitHub
      * Podrá enviar varios mensajes, el único que tendrá valor será el último

*  Recompensa:
     * Todos los integrantes de todos los grupos que logren un ordenamiento perfecto sumarán a su nota final el valor de 1.0 ,  siempre la nota máxima posible de toda la asignatura será de 10.0
     * Un ordenamiento que no sea el correcto, suma 0 puntos.
*  Será muy valioso el intercambio de ideas en Zulip channel  #CazaTalentos  topic  Intercambio de Ideas, este desafío suele ser muy polémico.
*  La participación en este desafío es opcional.


# Solucion

In [12]:
using Statistics

# asumimos el rate de la elegida
elegida = .8

# los pelotones estuvieron cerca, asumimos rates cercanos.
peloton_1 = .701:.001:0.799
peloton_2 = .700:.0005:0.799

# descomentar para comparar con peores jugadoras
#peloton_1 = .601:.002:0.799
#peloton_2 = .600:.001:0.799

# para el caso 4, la otra jugadora tuvo .75
peloton_4 = .75


# hacemos un vector con todas las jugadoras
jugadoras_1 = vcat(peloton_1, elegida)
# aqui aprovechamos que el peloton 1 tiene 99 y el peloton 100, para asi sumar 200 con la elegida.
jugadoras_2 = vcat(peloton_2, elegida)
jugadoras_4 = vcat(peloton_4, elegida)

# cada cazatalento tiene:
#[vector de jugadoras, cantidad de veces que gana la elegida, suma de diferencias de la ganadora en un segudo tiro]

cazatalentos= Dict(
  1 => [jugadoras_1, 0, 0],
  2 => [jugadoras_2, 0, 0],
  4 => [jugadoras_4, 0, 0]
)

n_experimentos = 50000

for caza in cazatalentos
  for i = 1:n_experimentos
    vaciertos  = ftirar.(caza[2][1], 100 )  # cada jugador tira 100 tiros libres
    _, mejor  = findmax( vaciertos )
    aciertos_torneo  = vaciertos[ mejor ]

    if mejor == length(caza[2][1])
      caza[2][2] += 1
    end

    aciertos_segunda  = ftirar.(caza[2][1][mejor], 100 )

    caza[2][3]  += (aciertos_torneo - aciertos_segunda )
  end
  println("Cazatalentos #$(caza[1])\tGana elegida: $(round(caza[2][2] / n_experimentos * 100, digits=2))%\tDiferencias: $(round(caza[2][3]/ n_experimentos,digits=4))" )
end



# Casos en los que la elegida por la cazatalentos 5 tire 80/100
caza_5_gte_80 = 0
promedio_caza5 = mean([68,74,78,70,68,63,80,68,67,65]) / 100
# loopeamos 100k
for i = 1:n_experimentos
  if ftirar(promedio_caza5,100) >= 0.8
    caza_5_gte_80 += 1
  end
end
print("Chances de que la elegida por la cazatalentos 5 tire 80/100: $(round(caza_5_gte_80/n_experimentos *100,digits=2))%")


LoadError: UndefVarError: `ftirar` not defined in `Main`
Suggestion: check for spelling errors or missing imports.

In [129]:
# Vamos a comparar la ganadora de cada método entre si, tirando 100 veces.

using Statistics

ftirar = (prob, q) -> (sum(rand(q) .< prob)/ q)


# asumimos rate de elegida
elegida = 0.8

# los pelotones estuvieron cerca, asumimos rates cercanos.
peloton_1 = .701:.001:0.799
peloton_2 = .700:.0005:0.799

# descomentar para comparar con peores jugadoras
#peloton_1 = .601:.002:0.799
#peloton_2 = .600:.001:0.799

# para el caso 4, la otra jugadora tuvo .75
peloton_4 = .75


# hacemos un vector con todas las jugadoras
jugadoras_1 = vcat(peloton_1, elegida)
# aqui aprovechamos que el peloton 1 tiene 99 y el peloton 100, para asi sumar 200 con la elegida.
jugadoras_2 = vcat(peloton_2, elegida)
jugadoras_4 = vcat(peloton_4, elegida)

cazatalentos = Dict(
  1 => [jugadoras_1],
  2 => [jugadoras_2],
  4 => [jugadoras_4]
)

# aqui vamos a guardar las ganadoras en cada experimento.
elegidas = Dict(
  1 => 0,
  2 => 0,
  3 => 0.8, # sin sesgo
  4 => 0,
  5 => 0.701 # media de sus tiros
)

# diccionario para guardar los resultados de un hipotetico matchup
wins_against = Dict(
  5 => Dict(4=>0, 3=>0, 2=>0, 1=>0),
  4 => Dict(3=>0, 2=>0, 1=>0),
  3 => Dict(2=>0, 1=>0),
  2 => Dict(1=>0),

)

function comparar_elegidas(elegidas, wins_against)
  # iteramos a traves de las ganadoras
  for elegida in (elegidas)
    for elegida_2 in (elegidas)
      # salteamos comparaciones ya hechas o con la elegida misma
      if elegida[1] <= elegida_2[1] continue end
      # comparamos el resultado de 100 tiros
      if ftirar(elegida[2], 100) > ftirar(elegida_2[2],100)
        # sumamos un win de la elegida frente al matchup
        wins_against[elegida[1]][elegida_2[1]] += 1
      end
    end
  end
end

n_experimentos = 10000

# simulamos 10000 selecciones
for i = 1:n_experimentos
  for caza in cazatalentos
    vaciertos  = ftirar.(caza[2][1], 100)  # cada jugadora tira 100 tiros libres
    _, mejor_index  = findmax(vaciertos) # Obtenemos el indice de la mejor jugadora

    elegidas[caza[1]] = caza[2][1][mejor_index] # asignamos ese rate al grupo de ganadoras para este cazatalentos

  end

  # en cada iteración comparamos las 5 elegidas
  comparar_elegidas(elegidas, wins_against)
end


# Diccionario para ver cuántos matchups gana cada elegida
final_scores = Dict(
  1=>0,
  2=>0,
  3=>0,
  4=>0,
  5=>0,
)

# cada key de wins_against es una elegida
for elegida in wins_against
  # cada valor de la elegida es el matchup con las restantes
  for matchup in elegida[2]
    # dividimos por la cantidad de experimentos para obtener el win rate.
    # matchup[1] es el # de elegida, matchup[2] la cantidad de wins
    if matchup[2] / n_experimentos < 0.5
      # aqui la elegida perdió mas del 50%, por lo que sumamos un score al matchup
      final_scores[matchup[1]] += 1
    else
      # aqui ganó más la elegida
      final_scores[elegida[1]] += 1
    end
  end
end

# ordenamos en base a los resultados
elegidas_ordenadas = sort(collect(final_scores), by = x -> x.second)
for elegida in elegidas_ordenadas
  print("C$(elegida[1]) < ")
end

C5 < C2 < C1 < C4 < C3 < 

In [90]:
ftirar = (prob, q) -> (sum(rand(q) .< prob)/ q)



comparar_ganadoras (generic function with 3 methods)

4 de los 5 cazatalentos realizaron el experimento de comparar N jugadoras tirando 100 tiros, siendo N:
100 para el cazatalentos 1
200 para el cazatalentos 2
1 para el cazatalentos 3
2 para el cazatalentos 4

En estos casos, si repetimos el experimento de tirar 100 tiros muchas veces (> 10000 veces), observo que para un peloton de similares caracteristicas, el cazatalentos 1 elije la jugadora con ratio 0.8 casi el doble de veces que el cazatalentos 2. De aqui consideramos que C2 < C1.

Para el cazatalentos 4, al tener solo 2 tiradoras, el efecto observado para los cazatalentos 1 y 2 no es tan fuerte. Es decir, es menos probable que se haya elegido a una tiradora cuyo rate no sea 0.8. Consideramos C2 < C1 < C4.

Siguiendo este analisis, para el cazatalentos #3 solo disponemos del dato de la unica tiradora, que demostro un promedio de 0.8. No aplica 'la maldicion de la ganadora' ni tenemos datos que indiquen que su verdadero promedio no sea 0.8. Considero entonces C2 < C1 < C4 < C3.

Finalmente, para el cazatalentos 5 solo se observan 10 tiradas de la misma jugadora, por lo que podemos considerar la media de los disparos como su valor esperado, 70.1. Corremos experimentos para ver que posibilidad hay que tire 80/100 con ese promedio, y el resultado es aproximadamente 1.6%.

Como conclusion, considero C5 < C2 < C1 < C4 < C3.


Chances de que la elegida por la cazatalentos 5 tire 80/100: 1.75%

0.74